File: Exercises/ExercisesFortran/08Jan_Chapter06/SingleChain/Source/mcloop.f
      Subroutine Mcloop
C     Generate multiple configurations of a polymer chain and calculate its averaged properties (energy, end-to-end distance) 
      Implicit None

      Include 'system.inc'

C     Local variables
      Integer I,J,K

      Double Precision Xst(Maxchain),Yst(Maxchain),
     &     Zst(Maxchain),Weight,Weiold,Ubnew,
     &     Ubold,Unnew,Unold,R,RandomNumber,
     &     Bn,Bs,AvUbond,AvUnbond,NoConfs,Eb,En

C     Bn:              number of attempts to create a chain
C     Bs:              number of accepted chains 
C     AvUbond:         average bonded energy
C     AvUnbond:        average non bonded energy
C     NoConfs:         number of configurations used to compute averages 
C     Ubold, Ubnew:    bonded energy of the old and trial chains
C     Unoldn Unnew:    non bonded energy of the old and trial chains
C     Weiold, Weight:  statistical weight of the old and trial chains.  If random chains are generated, both values are 1.
C     R:               end-to-end distance
C     Eb,En:           Bonded and non bonded energy of the  chain 

C     initiallize variables
      Weight = 1.0d0
      Ubnew  = 0.0d0
      Unnew  = 0.0d0
      Weiold = 1.0d0
      Ubold  = 0.0d0
      Unold  = 0.0d0
      Bn     = 0.0d0
      Bs     = 0.0d0
      AvUbond    = 0.0d0
      AvUnbond    = 0.0d0
      NoConfs    = 0.0d0
      
C     Generate the very first Chain  
      Call Grow(.False.,Weiold,Ubold,Unold,Xst,Yst,Zst)

      Do J=1,ChainLength
         Xpos(J) = Xst(J)
         Ypos(J) = Yst(J)
         Zpos(J) = Zst(J)
      Enddo

      Eb = Ubold
      En = Unold

C     Initialize the subroutine "Sample" (the subroutine that will sample the average end-to-end distance)
      Call Sample(1,1.0d0,1.0d0)

C     Generate the Markov chain of states 
      Do I=1,NumberOfSteps
         if (MOD(I,20) .EQ. 0) THEN
           print*,"cycle",I
         ENDIF
         Do K = 1,1000

               Bn = Bn + 1.0d0

C              retrace old chain 
               Call Grow(.True. ,Weiold,Ubold,Unold,Xst,Yst,Zst)

C              create a new chain
               Call Grow(.False.,Weight,Ubnew,Unnew,Xst,Yst,Zst)

C              If the new chain is accepted, then Copy Coordinates   
               If(RandomNumber().Lt.(Weight/Weiold)) Then

                  Bs = Bs + 1.0d0
                  En = Unnew
                  Eb = Ubnew

                  Do J=1,ChainLength
                     Xpos(J) = Xst(J)
                     Ypos(J) = Yst(J)
                     Zpos(J) = Zst(J)
                  Enddo
               Endif

               Weight = 1.0d0

C           If the simulation is beyond the set number of equilibration steps,
C           sample End-To-End Distance And Average Energies     
            If(I.Gt.NumberOfInitializationSteps) Then
               R = Dsqrt((Xpos(ChainLength)-Xpos(1))**2 +
     &                 (Ypos(ChainLength)-Ypos(1))**2 +
     &                 (Zpos(ChainLength)-Zpos(1))**2)
               Call Sample(2,R,Weight)
               AvUbond = AvUbond + Eb
               AvUnbond = AvUnbond + En

C              update number of configurations sampled  
               NoConfs = NoConfs + 1.0d0

C              Periodically write a pdb file containing the particle coordinates 
               If(K.Eq.1.And.Mod(I,5).Eq.0) Call Writepdb
            Endif
         Enddo
      Enddo

C     Call subroutine "Sample" to output the histogram of the end-to-end distance
      Call Sample(3,1.0d0,1.0d0)

C    Output average energies and fraction of accepted chains 
      Write(6,*) 'Fraction Accepted : ',Bs/Bn
      Write(6,*) 'Average U-Bonded  : ',AvUbond/NoConfs
      Write(6,*) 'Average U-Nonb    : ',AvUnbond/NoConfs

      Return
      End

File: Exercises/ExercisesC/08Jan_Chapter06/SingleChain/Source/mcloop.c
// Generate multiple configurations of a polymer chain and calculate its averaged properties (energy, end-to-end distance)
#include <stdlib.h>
#include <stdio.h>
#include <math.h>
#include "system.h"
#include "ran_uniform.h"

void Mcloop(void)
{
  int i,j,k;
  int NumberAccepted,NumberTrials;
  double RosenbluthFactor,RosenbluthFactorOld,r;
  double BondedEnergyOld,BondedEnergyNew,NonBondedEnergyNew,NonBondedEnergyOld;
  double BondedEnergySum,NonBondedEnergySum,EnergyNormalize,BondedEnergy,NonBondedEnergy;
  FILE *FilePtr;

  FilePtr=fopen("movie.pdb","w");
  InitializePdb(FilePtr);

//   NumberTrials:                               number of attempts to create a chain
//   NumberAccepted:                             number of accepted chains 
//   BondedEnergySum:                            average bonded energy
//   NonBondedEnergySum:                         average non bonded energy
//   EnergyNormalize:                            number of configurations used to compute averages 
//   BondedEnergyOld, BondedEnergyNew:           bonded energy of the old and trial chains
//   NonBondedEnergyOld, NonBondedEnergyNew:    non bonded energy of the old and trial chains
//   RosenbluthFactorOld, RosenbluthFactor:      statistical weight of the old and trial chains.  If random chains are generated, both values are 1.
//   r:                                          end-to-end distance
//   BondedEnergy,NonBondedEnergy:                Bonded and non bonded energy of the  chain 

  // initiallize variables
  RosenbluthFactor=1.0;
  BondedEnergyNew=0.0;
  NonBondedEnergyNew=0.0;
  RosenbluthFactorOld=1.0;
  BondedEnergyOld=0.0;
  NonBondedEnergyOld=0.0;
  NumberTrials=0.0;
  NumberAccepted=0.0;
  BondedEnergySum=0.0;
  NonBondedEnergySum=0.0;
  EnergyNormalize=0.0;
      
  // generate a first chain
  Grow(FALSE,&RosenbluthFactorOld,&BondedEnergyOld,&NonBondedEnergyOld);


  for(j=0;j<ChainLength;j++)
  {
    Positions[j].x=TrialPositions[j].x;
    Positions[j].y=TrialPositions[j].y;
    Positions[j].z=TrialPositions[j].z;
  }

  BondedEnergy=BondedEnergyOld;
  NonBondedEnergy=NonBondedEnergyOld;

  // Initialize the subroutine "Sample" (the subroutine that will sample the average end-to-end distance)
  Sample(INITIALIZE,1.0,1.0);

  // Generate the Markov chain of states
  for(i=0;i<NumberOfSteps;i++)
  {
    if(fmod(i,20) == 0) {
      printf("cycle: %d \n", i);
    }
    for(k=0;k<1000;k++)
    {
      NumberTrials++;
      // retrace old chain
      Grow(TRUE,&RosenbluthFactorOld,&BondedEnergyOld,&NonBondedEnergyOld);
      // create a new chain
      Grow(FALSE,&RosenbluthFactor,&BondedEnergyNew,&NonBondedEnergyNew);

      // If the new chain is accepted, then copy coordinates
      if(RandomNumber()<(RosenbluthFactor/RosenbluthFactorOld))
      {
        NumberAccepted++;
        NonBondedEnergy=NonBondedEnergyNew;
        BondedEnergy=BondedEnergyNew;

        for(j=0;j<ChainLength;j++)
        {
          Positions[j].x=TrialPositions[j].x;
          Positions[j].y=TrialPositions[j].y;
          Positions[j].z=TrialPositions[j].z;
        }
      }
      RosenbluthFactor=1.0;

      // If the simulation is beyond the set number of equilibration steps,
      // sample End-To-End Distance And Average Energies
      if(i>=NumberOfInitializationSteps) 
      {
        r=sqrt(SQR(Positions[ChainLength-1].x-Positions[0].x)+
              SQR(Positions[ChainLength-1].y-Positions[0].y)+
              SQR(Positions[ChainLength-1].z-Positions[0].z));
        Sample(SAMPLE,r,RosenbluthFactor);
        BondedEnergySum+=BondedEnergy;
        NonBondedEnergySum+=NonBondedEnergy;

        // update number of configurations sampled
        EnergyNormalize+=1.0;

        // Periodically write a pdb file containing the particle coordinates
        if(k==0&&(i%5)==0) 
          WritePdb(FilePtr);
      }
    }
  }

  // Call subroutine "Sample" to output the histogram of the end-to-end distance
  Sample(WRITE_RESULTS,1.0,1.0);

  // Output average energies and fraction of accepted chains
  printf("Fraction Accepted           : %f\n", (double)NumberAccepted/(double)NumberTrials ); 
  printf("Average bonded energy       : %f\n",BondedEnergySum/EnergyNormalize);
  printf("Average nonbonded energy    : %f\n",NonBondedEnergySum/EnergyNormalize);

  fclose(FilePtr);
}
